tremor-script/src/srs.rs

```diff
@@ -0,0 +1,726 @@
+        Self {
+            raw: Vec::new(),
+            structured: vm.into(),
+        }
+    }
+}
+
+impl PartialEq for EventPayload {
+    fn eq(&self, other: &Self) -> bool {
+        self.structured.eq(&other.structured)
+    }
+}
+
+impl simd_json_derive::Serialize for EventPayload {
+    fn json_write<W>(&self, writer: &mut W) -> std::io::Result<()>
+    where
+        W: std::io::Write,
+    {
+        self.rent(|d| d.json_write(writer))
+    }
+}
+
+impl<'input> simd_json_derive::Deserialize<'input> for EventPayload {
+    fn from_tape(tape: &mut simd_json_derive::Tape<'input>) -> simd_json::Result<Self>
+    where
+        Self: Sized + 'input,
+    {
+        let ValueAndMeta { v, m } = simd_json_derive::Deserialize::from_tape(tape)?;
+
+        Ok(Self::new(vec![], |_| {
+            ValueAndMeta::from_parts(v.into_static(), m.into_static())
+        }))
+    }
+}
+
+/*
+=========================================================================
+*/
+
+/// Combined struct for an event value and metadata
+#[derive(
+    Clone, Debug, PartialEq, Serialize, simd_json_derive::Serialize, simd_json_derive::Deserialize,
+)]
+pub struct ValueAndMeta<'event> {
+    v: Value<'event>,
+    m: Value<'event>,
+}
+
+impl<'event> ValueAndMeta<'event> {
+    /// A value from it's parts
+    #[must_use]
+    pub fn from_parts(v: Value<'event>, m: Value<'event>) -> Self {
+        Self { v, m }
+    }
+    /// Event value
+    #[must_use]
+    pub fn value(&self) -> &Value<'event> {
+        &self.v
+    }
+
+    /// Event value
+    #[must_use]
+    pub fn value_mut(&mut self) -> &mut Value<'event> {
+        &mut self.v
+    }
+    /// Event metadata
+    #[must_use]
+    pub fn meta(&self) -> &Value<'event> {
+        &self.m
+    }
+    /// Deconstruicts the value into it's parts
+    #[must_use]
+    pub fn into_parts(self) -> (Value<'event>, Value<'event>) {
+        (self.v, self.m)
+    }
+    /// borrows both parts as mutalbe
+    #[must_use]
+    pub fn parts_mut(&mut self) -> (&mut Value<'event>, &mut Value<'event>) {
+        (&mut self.v, &mut self.m)
+    }
+
+    /// borrows both parts as mutalbe
+    #[must_use]
+    pub fn parts(&self) -> (&Value<'event>, &Value<'event>) {
+        (&self.v, &self.m)
+    }
+}
+
+impl<'event> Default for ValueAndMeta<'event> {
+    fn default() -> Self {
+        ValueAndMeta {
+            v: Value::object(),
+            m: Value::object(),
+        }
+    }
+}
+
+impl<'v> From<Value<'v>> for ValueAndMeta<'v> {
+    fn from(v: Value<'v>) -> ValueAndMeta<'v> {
+        ValueAndMeta {
+            v,
+            m: Value::object(),
+        }
+    }
+}
+
+impl<'v, T1, T2> From<(T1, T2)> for ValueAndMeta<'v>
+where
+    Value<'v>: From<T1> + From<T2>,
+{
+    fn from((v, m): (T1, T2)) -> Self {
+        ValueAndMeta {
+            v: Value::from(v),
+            m: Value::from(m),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+
+    // /// this is commented out since it should fail
```

Comment:
you could move this to a doctest using `compile_fail` like so:

```rust
/// ```compile_fail
/// let x = 5;
/// x += 2; // shouldn't compile!
/// ```
```

Reply:
I like that!